### PR TITLE
Api bzz react native

### PR DESCRIPTION
**Pull request by vujevits**
### Changes
Implements partial support for react-native, for the `bzz` functionality in particular. Related issue #52 
There are 2 differences compared to `api-bzz-browser`: 
- the `URL` type is used for other purposes in react-native, so in the implementation we have to depend on a URL type provided by `universal-url`
- `Blob` and `FormData` has some differences in react-native and typing issues (both in flow and TypeScript), this had to be resolved as well

### Todo
Run tests in detox, or other method which is closer to react-native then node / browser.

This is a draft PR, I opened it to make further discussions easier. 

I had some tests failing locally:
 FAIL  __tests__/timeline.js
  ● Test suite failed to run

 FAIL  __tests__/api-bzz-node.js
  ● api-bzz-node › encountered a declaration exception

    ReferenceError: URL is not defined

### CODE REVIEW

**Comment by PaulLeCam** [comment]
Thanks for your PR!
Regarding the errors running tests, are you using node v10+ please? I think older versions don't expose the `URL` class globally.

**Comment by vujevits** [comment]
damn nvm :) I was running the tests on another tab with v8.9.0. 
Btw I noticed that there are commands which exit with an error if you are using node <v10, but some aren't.

**Comment by PaulLeCam** [comment]
I think some tests probably have dangling timers (notably when testing subscriptions) that aren't cleared when an error is thrown, I'll try to go through these when I have the chance.

**Comment by vujevits** [comment]
Hi, I was thinking about how to move forward. I can create a simple react-native application, where we can run functionally the same tests as in the browser test suite, I'm just wondering what would be the best setup from the repository and the CI/CD point of view. I can imagine the following setups:
- the react-native test up is a package in this repo, making any changes to the repo would trigger running those tests as well
- the react-native app and the tests are in a separate repo, and the `erebos` build should trigger its test-suite
- the react-native app is in a sepearate repo, the tests are in `erebos`, and somehow we load it in the react-native app

The first one is the most straightforward, but I'm not sure if you want to put a relatively heavy sub-project (the most basic react-native app) to this repo.
The second one needs some additional work regarding CI/CD, but I'm not sure about Mainframe's setup, maybe it's something you already are doing with some other projects. Another downside to consider is that the tests are in a separate repo, which should be updated when there is new functionality to `erebos`.
The third one might be technically difficult, without further investigation I'm not sure if it can be achieved easily.

**Comment by PaulLeCam** [comment]
Hey, thanks for looking into this!

I agree that the first option is likely the simplest setup.
I don't mind adding the react-native dependencies in the repository if it's needed for tests anyways, as long as it doesn't prevent installation on any platform (so running `yarn install` on Windows should work even if iOS tests couldn't be run on Windows).

For CI currently we're using Circle but it doesn't have a free macOS plan (which I assume is needed to test iOS?), so I'm thinking about switching to Travis or Azure Pipelines, do you have any recommendation please?
Thanks!

**Comment by vujevits** [comment]
We are using Travis, sometimes we saw XCode builds fail in a non-reproducible way, but it was rare, and haven't seen it in a while either. I'll have a look at creating the app in the coming weeks.

**Comment by PaulLeCam** [comment]
Great, thanks!
I'll try to move our current CI setup from Circle to Travis if I have time this week, it should simplify the process.

Also FYI I started rewriting Erebos in TypeScript so it may affect your changes, here is the current changelog: https://github.com/MainframeHQ/erebos/blob/typescript/CHANGELOG.md#v090-unreleased
I was planning on releasing v0.9 with this rewrite once `typescript-eslint` v2 is released with `eslint` v6 support, but if support for React-Native is around the corner I think it makes sense to include it in v0.9 as well, what do you think?

**Comment by vujevits** [comment]
Sure! Happy to hear you are moving to TypeScript!

**Comment by PaulLeCam** [comment]
Thanks!
I added the Travis config to https://github.com/MainframeHQ/erebos/blob/master/.travis.yml please feel free to edit as needed to make it work for React-Native integration tests!

File: packages/api-bzz-react-native/package.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "build": "yarn clean && yarn build:js && yarn build:types",
+    "prepublishOnly": "yarn build"
+  },
+  "dependencies": {
+    "@babel/runtime": "^7.5.5",
+    "@erebos/api-bzz-base": "^0.9.0",
+    "universal-url": "^2.0.0"
+  },
+  "devDependencies": {
+    "flow-bin": "^0.101.0"
```
Comment (PaulLeCam) [inline packages/api-bzz-react-native/package.json]: Could you remove this dependency now that the project uses TypeScript please?

File: packages/api-bzz-react-native/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  UploadOptions,
+} from '@erebos/api-bzz-base'
+import { hexValue } from '@erebos/hex'
+
+export * from '@erebos/api-bzz-base'
+
+export class Bzz extends BaseBzz<Response> {
+  public constructor(config: BzzConfig) {
+    const { url, ...cfg } = config
+    super(window.fetch.bind(window), { ...cfg, url: new URL(url).href })
```
Comment (PaulLeCam) [inline packages/api-bzz-react-native/src/index.ts]: Shouldn't this use `global` instead of `window`?

File: packages/timeline/types/index.d.ts [hunk 3]
```diff
@@ -1,69 +0,0 @@
-/// <reference types="node" />
```
Comment (PaulLeCam) [inline packages/timeline/types/index.d.ts]: Any reason this file is deleted please?

File: yarn.lock [hunk 4]
```diff
@@ -4427,6 +4427,11 @@ flatted@^2.0.0:
   resolved "https://registry.yarnpkg.com/flatted/-/flatted-2.0.1.tgz#69e57caa8f0eacbc281d2e2cb458d46fdb449e08"
   integrity sha512-a1hQMktqW9Nmqr5aktAux3JMNqaucxGcjtjWnZLHX7yyPCmlSV3M54nGYbqT8K+0GhF3NBgmJCc3ma+WOgX8Jg==
 
+flow-bin@^0.101.0:
```
Comment (PaulLeCam) [inline yarn.lock]: Could you make sure this get removed once `flow-bin` has been removed from the `package.json` please?

File: packages/api-bzz-react-native/src/index.ts (hunk 2 above)
Comment (vujevits) [reply packages/api-bzz-react-native/src/index.ts]: Not sure if this is needed, fetch is available globally in react-native, should we just pass in `fetch`?

Comment (PaulLeCam) [reply packages/api-bzz-react-native/src/index.ts]: I'm not sure in RN but on the Web it needed to be bound to the proper context, there was some issue without.

**Comment by PaulLeCam** [comment]
Thanks!

**Pull request merged**